src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,168 @@
+                notTheHeadAnyMore.prev = head;
+                size++;
+            } else {
+                int currentIndex = 0;
+                Node currentNode = head;
+                while (currentIndex != index) {
+                    currentNode = currentNode.next;
+                    currentIndex++;
+                }
+                Node newNode = new Node(currentNode.prev, value, currentNode);
+                currentNode.prev.next = newNode;
+                currentNode.prev = newNode;
+
+                size++;
+            }
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new IndexOutOfBoundsException("given index does not exist");
+        } else {
+            int currentIndex = 0;
+            Node currentNode = head;
+            while (currentIndex != index) {
+                currentNode = currentNode.next;
+                currentIndex++;
+            }
+            return (T) currentNode.value;
+        }
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new IndexOutOfBoundsException("given index does not exist");
+        } else {
+            int currentIndex = 0;
+            Node currentNode = head;
+            while (currentIndex != index) {
+                currentNode = currentNode.next;
+                currentIndex++;
+            }
+            T previousNode = (T) currentNode.value;
+            currentNode.value = value;
+            return previousNode;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new IndexOutOfBoundsException("given index does not exist");
+        } else if (size == 1) {
+            Node removedNode = head;
+            head = null;
+            size--;
+            return (T) removedNode.value;
+        } else if (index == 0) {
+            Node removedNode = head;
+            head.next.prev = null;
+            head = head.next;
+            size--;
+            if (removedNode == null) {
+                return null;
+            }
+            return (T) removedNode.value;
+        } else if (index == size - 1) {
+            Node removedNode = tail;
+            tail.prev.next = null;
+            tail = tail.prev;
+            size--;
+            return (T) removedNode.value;
+        } else {
+            int currentIndex = 0;
+            Node currentNode = head;
+            while (currentIndex != index) {
+                currentNode = currentNode.next;
+                currentIndex++;
+            }
+            Node removedNode = currentNode;
+            currentNode.prev.next = currentNode.next;
```

Comment:
Create a separate method `unlink(Node node)`. This logic will be used in 2 methods: when we remove from the list by index and by value. So let's create a method that will take a Node that needs to be removed and unlink it. As a result, we will call it from both methods when Node is found.